Observability integration interop testing - first draft

stanley-cheung opened the pull request:
GCP Observability Integration Interop Testing - first draft

Key files to review:
- [`build.sh`](18/files#diff-81abf1610e704a0cf3f7c9dbad347a31f8ca76266d828de6132c74188905b88d): (~150 lines) the entry shell script to be run by each Kokoro job. This bash script installs pre-requisites packages in the Kokoro VM, builds the binaries for each supported language and launches the main test driver (.py), which is the following
- [`o11y_tests_manager.py`](18/files#diff-6b30c52e9732cd0a5587f9ab63870f5620d8caaed88e58e09aa110506891290b): (~300 lines) this script decides how many parallel test queues to run and launches each individual test case as separate processes (see the next script)
- [`run_o11y_tests.py`](18/files#diff-0c8025ab3d7d64f3d7e1b856da6e4f5e0b825959ffa6afaf221a79475262f2db): (~1200 lines) this script runs _one_ test case and contains the majority of the business logic of the testing framework. A typical test case involves
  - Setting up the GCP observability config, according to the test spec
  - Starting a server (in one language)
  - Running a client action (could be the same or different language), i.e. send an RPC
  - Querying Cloud Logging / Monitoring / Trace to see if the expected data are produced
- [`test_utils.py`](18/files#diff-64870e8305b7267c0a954a30f529f3e6981e16f0248705519448431cfbd9ffce): (~170 lines) this contains some common code and types shared by multiple .py scripts

gnossen commented:
@sergiitk would also be a great reviewer because of his experience with the PSM interop.

ejona86 commented on observability/test/run_o11y_tests.py:
```diff
@@ -0,0 +1,1174 @@
+LANGUAGE_CONSTANTS = {
+    SupportedLangEnum.GO: LanguageConstants(
+        repo_name = 'grpc-go',
+        server_start_cmd = 'interop/observability/server/server',
```
Please let's not have custom configuration for each. We should define the arguments including choosing between client/server and all support that specific set.

stanley-cheung replied:
Yes I plan to eventually remove this and make them parameters (or just a uniform naming scheme somewhere). I will either add a TODO (more likely) or fix it by the next revision (less likely).

stanley-cheung replied:
This whole `LANGUAGE_CONSTANTS` block has been removed! Now these details live in individual language's `Dockerfile` and `run.sh`

ejona86 commented on observability/test/run_o11y_tests.py:
```diff
@@ -0,0 +1,1174 @@
+            return m.group(1)
+        return False
+
+    def exec_commands_in_pod(self,
```
Why did you go with this approach? I'd much rather we pass args to the container when it starts.

ejona86 replied:
Oh, I didn't notice you weren't using docker locally when running locally. It should really be docker there, and we get rid of the `sleep inf` goo.

stanley-cheung replied:
This has been implemented.

Now individual language needs to supply a `Dockerfile` that has a `run.sh` `ENTRYPOINT` which takes a pre-defined set of parameters to determine how the interop client/server is executed.

This allows this main test driver to just execute `docker run <docker image> <additional parameters>` and don't need to worry about knowing individual language's implementation.

gnossen commented on .gitignore:
```diff
@@ -0,0 +1,3 @@
+py_compile.sh
```
Are these generated files? If not, you probably want to put them in your personal, global gitignore, rather than this project-specific one.

stanley-cheung replied:
Yes my mistake.

stanley-cheung replied:
This file has been removed.

gnossen commented on build/kokoro/build.sh:
```diff
@@ -0,0 +1,163 @@
+python3 -m pip install -r requirements.txt
+
+# Install Go
+curl -sSL https://go.dev/dl/go1.17.13.linux-amd64.tar.gz -o /tmp/go1.17.13.linux-amd64.tar.gz
```
This download should be verified with a checksum.

stanley-cheung replied:
This has been added.

gnossen commented on build/kokoro/build.sh:
```diff
@@ -0,0 +1,163 @@
+
+# Install Python packages
+python3 --version
+python3 -m pip install -r requirements.txt
```
I would recommend splitting this into two files:
 - `requirements.txt` with version ranges specified, but not specific versions
 - `requirements.lock` generated from `requirements.txt` with specific versions specified. These are the versions actually used in test runs.

We have seen several PSM interop outages caused by pushes of new versions of Python packages. By freezing to particular packages, you'll protect yourself from non-hermetic breakage (and a lot of headaches).

stanley-cheung replied:
This has been implemented. Added a `requirements.lock` file generated from running `pip freeze` in a virtualenv.

gnossen commented on observability/test/run_o11y_tests.py:
```diff
@@ -0,0 +1,1174 @@
+class GKETestEnv:
+    args: argparse.Namespace
+    k8s_core_v1: kubernetes_client.CoreV1Api
+    k8s_client: kubernetes_client.ApiClient
```
We just learned an interesting quirk about this library that you'll want to incorporate into your code: https://github.com/grpc/grpc/pull/32210

You should only see an issue with this however if:

1. The tests exceed ~45 minutes
2. or the GCP token lifetime is decreased on the backend

Neither of these is likely to happen immediately, but it will be real headache if/when it does.

stanley-cheung replied:
I had made a note of this. Thanks for the pointer.

gnossen commented on build/kokoro/build.sh:
```diff
@@ -0,0 +1,163 @@
+##
+
+build_java () {
+  mkdir -p ${REPOS_BASE_DIR}/grpc-java
```
In general, I would recommend doing builds within docker images for hermeticity. Kokoro images can change underneath you.

stanley-cheung replied:
This has been implemented. Most of the build instructions have been done in individual language's `Dockerfile`. There is still a call out to the entry point script in individual language repo here. I plan to improve that in the future.

gnossen commented on observability/test/manual_deploy.sh:
```diff
@@ -0,0 +1,12 @@
+#!/bin/bash
```
One thing I don't see in this PR is documentation. It will take a lot of pressure off of you to be the only person who can operate the interop framework if you document:

- what the most important files _do_
- how to run them locally
- if you _can't_ easily run locally, how to set up an env in which you _can_ run them

Take a look at [this](https://github.com/grpc/grpc/blob/master/tools/run_tests/xds_k8s_test_driver/README.md) for an example

stanley-cheung replied:
Yes I plan to add those instructions in the next set of PRs. One of the reasons I haven't done it so far is that _the way things work_ have been changing a lot. So I had wanted things to stabilize a bit before adding those instructions thinking they are going to get completely overhauled anyways.

gnossen commented on observability/test/gke-deployment.yaml:
```diff
@@ -0,0 +1,151 @@
+apiVersion: v1
+kind: Namespace
+metadata:
+  name: grpc-server-ns-${JOB_MODE}
```
You probably want to put some unique ID in these names so that you can have multiple test runs going in parallel.

stanley-cheung replied:
1. I had removed all of the GKE related code for this iteration for now, because now individual language's interop client/server docker containers are now being run by `docker run` command, things need to be changed on the GKE deployment side. For that change I find it just too much for this PR so I will deal with the GKE aspect of things in a separate PR.  For now, this PR will only execute all the tests in the Kokoro VM.

2. Yes I added a TODO to add a unique identifier to each docker container we just built to make sure each test run is running against that exact docker container just being built.

gnossen commented on observability/test/manual_deploy.sh:
```diff
@@ -0,0 +1,12 @@
+
+export PROJECT=`gcloud config get-value project`
+export PROJNUM=`gcloud projects describe $PROJECT --format="value(projectNumber)"`
+export SERVER_IMAGE=gcr.io/microsvcs-testing/grpc-observability/testing/integration-dev-go:1.53.0-dev
```
I see multipole references to 1.53 in this PR. We should probably make these references to a single place where the gRPC version is defined.

stanley-cheung replied:
Yes I added a TODO to hopefully remove the need for this constant.

gnossen commented on observability/test/manual_deploy.sh:
```diff
@@ -0,0 +1,12 @@
+export PROJECT=`gcloud config get-value project`
+export PROJNUM=`gcloud projects describe $PROJECT --format="value(projectNumber)"`
+export SERVER_IMAGE=gcr.io/microsvcs-testing/grpc-observability/testing/integration-dev-go:1.53.0-dev
+export CLIENT_IMAGE=gcr.io/microsvcs-testing/grpc-observability/testing/integration-dev-go:1.53.0-dev
```
Please make sure that multiple other people on the team have access to this project.

stanley-cheung replied:
Yes most of the gcp observability team is on that project.

gnossen commented on observability/test/o11y_tests_manager.py:
```diff
@@ -0,0 +1,306 @@
+            thd.join()
+            logger.info('Queue %s is done' % queue_key)
+
+# Main
```
Nit: include with a block like the following:

```python
if __name__ == "__main__":
    test_runner = TestRunner(parse_args())
    test_runner.run_tests()
```

stanley-cheung replied:
Implemented.

gnossen commented on observability/test/o11y_tests_manager.py:
```diff
@@ -0,0 +1,306 @@
+            env = os.environ.copy()
+            env['GRPC_VERBOSITY'] = 'DEBUG'
+            return subprocess.Popen(
+                [os.path.join(os.path.dirname(__file__), 'run_o11y_tests.py'),
```
Nit: Instead of relying on this file being executable an defining a shebang, it might be better to use [`sys.executable`](https://docs.python.org/3/library/sys.html#sys.executable). This might avoid potential mismatches between desired Python binary used and the actual binary used.

stanley-cheung replied:
Implemented.

gnossen commented on observability/test/o11y_tests_manager.py:
```diff
@@ -0,0 +1,306 @@
+            logger.info('Starting %s, in queue %s' % (job.job_name, queue_key))
+            env = os.environ.copy()
+            env['GRPC_VERBOSITY'] = 'DEBUG'
+            return subprocess.Popen(
```
Nice job! Folks almost never get subprocesses quite right in Python. :+1:

gnossen commented on observability/test/test_utils.py:
```diff
@@ -0,0 +1,177 @@
+        artifacts_dir = os.environ.get('KOKORO_ARTIFACTS_DIR')
+        if not artifacts_dir:
+            raise ValueError('Env var KOKORO_ARTIFACTS_DIR is not defined')
+        sponge_log_dir = os.path.join(artifacts_dir, str(job_mode), job_name)
```
Nit: Why the `str(job_mode)`? Isn't it already a `str`?

stanley-cheung replied:
Fixed.

gnossen reviewed (commented):
Not a full review, but I wanted to get a first batch of comments out you sooner rather than later.

gnossen reviewed (commented):
Another round. Still not a complete review.

General note: Please view all comments on optional. My goal is to help avoid future pitfalls, not to block the launch. I think what you've got here is already a very strong start.

gnossen commented on observability/test/test_utils.py:
```diff
@@ -0,0 +1,177 @@
+class TestUtil:
+    @staticmethod
+    def get_sponge_log_dir(job_mode: JobMode, job_name: str) -> str:
+        artifacts_dir = os.environ.get('KOKORO_ARTIFACTS_DIR')
```
:+1: Nice.

gnossen commented on observability/test/run_o11y_tests.py:
```diff
@@ -0,0 +1,1174 @@
+            logger.info('Killing server...')
+            os.kill(server_proc.pid, signal.SIGTERM)
+            server_proc.wait(timeout=5)
+            self.sponge_log_out.close()
```
Maybe send `SIGKILL` in this case.

stanley-cheung replied:
Implemented.

gnossen commented on observability/test/run_o11y_tests.py:
```diff
@@ -0,0 +1,1174 @@
+            logger.error(traceback.format_exc())
+        finally:
+            logger.info('Killing server...')
+            os.kill(server_proc.pid, signal.SIGTERM)
```
You should probably check the return code and print it out. Segfaults are possible and we need to know when they happen.

gnossen replied:
You're not also going to send `SIGTERM` to the client?

stanley-cheung replied:
So for now, the client is supposed to exit by itself after executing those actions, and flushing the observability logging/metrics/trace to cloud. Will see if this needs to be improved.

gnossen commented on observability/test/run_o11y_tests.py:
```diff
@@ -0,0 +1,1174 @@
+            server_env[CONFIG_ENV_VAR_NAME] = server_config_into_env_var.toJson()
+        if client_config_into_env_var:
+            client_env[CONFIG_ENV_VAR_NAME] = client_config_into_env_var.toJson()
+        server_proc = self.start_server_in_subprocess(env = server_env)
```
You should add a timestamped log statement every time you:

- Make a remote API call
- Start a subprocess
- A subprocess finishes
- You send a signal to a subprocess

This sort of stuff is crucial for debuggability.

stanley-cheung replied:
Noted. Agreed.

gnossen commented on observability/test/run_o11y_tests.py:
```diff
@@ -0,0 +1,1174 @@
+            gke_env.clean_up_gke_resources()
+            self.sponge_log_out.close()
+
+    def setup_and_run_rpc_vm(self,
```
Why the `_vm` part of the method name? I don't see a VM being set up here.

stanley-cheung replied:
Fixed. Now it's just `setup_and_run_rpc`.

gnossen commented on observability/test/run_o11y_tests.py:
```diff
@@ -0,0 +1,1174 @@
+        server_proc = self.start_server_in_subprocess(env = server_env)
+        time.sleep(WAIT_SECS_SERVER_START)
+        try:
+            client_proc = self.start_client_in_subprocess(action, env = client_env)
```
We talked about this previously, but I think this should probably move to a container-based test harness rather than a subprocess-based test harness. This is essential for Python and will make maintenance for other languages much easier as well.

stanley-cheung replied:
Implemented. Now individual language's interop client/server are container based.

gnossen commented on observability/test/run_o11y_tests.py:
```diff
@@ -0,0 +1,1174 @@
+CloudMonitoringInterfaceT = TypeVar('CloudMonitoringInterfaceT', bound='CloudMonitoringInterface')
+CloudTraceInterfaceT = TypeVar('CloudTraceInterfaceT', bound='CloudTraceInterface')
+
+class CloudLoggingInterface(unittest.TestCase):
```
The structure of this doesn't seem quite right. I would have expected `TestCaseImpl` to inherit from `unittest.TestCase`, not this helper class. This class defines assertions, not test cases.

stanley-cheung replied:
Agreed. Will try to correct the structure in the next round.

gnossen commented on observability/test/run_o11y_tests.py:
```diff
@@ -0,0 +1,1174 @@
+    results: List[LogEntry]
+
+    @classmethod
+    def query_logging_entries_from_cloud(cls: Type[CloudLoggingInterfaceT],
```
Couldn't this just be `Type[CloudLoggingInterface]` rather than `Type[CloudLoggingInterfaceT]`? I don't think any of the TypeVars above are actually necessary.

stanley-cheung replied:
Ditto from above.

stanley-cheung commented on build/kokoro/requirements-dev.txt:
```diff
@@ -0,0 +1,2 @@
+-r requirements.txt
+mypy~=0.991
```
Huge thanks to @gnossen for giving this PR a thorough and helpful review!

stanley-cheung commented on observability/test/o11y_tests_manager.py:
```diff
@@ -0,0 +1,306 @@
+        self.job_name = job_name
+        self.port = port
+
+class _BaseKey:
```
Fixed. Removed those 2 classes.

stanley-cheung commented on observability/test/o11y_tests_manager.py:
```diff
@@ -0,0 +1,306 @@
+    def initialize_single_lang(self, lang: SupportedLang) -> None:
+        self.curr_lang_shard_nums[lang] = 0
+        logger.info('Initializing %d job queues for %s' % (NUM_LANG_QUEUES, lang))
+        for i in range(0, NUM_LANG_QUEUES):
```
Fixed.

stanley-cheung commented on observability/test/o11y_tests_manager.py:
```diff
@@ -0,0 +1,306 @@
+class TestManager:
+    args: argparse.Namespace
+    test_runner: TestRunner
+    job_queues: Dict[QueueKey, JobQueue]
```
Fixed. Removed those 2 classes.

stanley-cheung commented on observability/test/o11y_tests_manager.py:
```diff
@@ -0,0 +1,306 @@
+        logger.info('Initializing %d job queues for each interop combinations' % NUM_INTEROP_QUEUES)
+        for entry in INTEROP_COMBINATIONS:
+            entry_key = InteropEntryKey('%s:%s' % (entry['server_lang'], entry['client_lang']))
+            self.curr_interop_shard_nums[entry_key] = 0
```
Fixed.

stanley-cheung commented on observability/test/test_utils.py:
```diff
@@ -0,0 +1,177 @@
+    def __str__(self) -> str:
+        return self.value
+
+class SupportedLang():
```
I added a TODO to try to simplify this. I think one of my original intention is to turn the input string (from command line args) into the Enum. And I might have things confused. I will try to fix this in the next round.

stanley-cheung commented on observability/test/o11y_tests_manager.py:
```diff
@@ -0,0 +1,306 @@
+        threads: Dict[QueueKey, threading.Thread] = {}
+        max_job_queue_size = 0
+        for queue_key, job_queue in self.job_queues.items():
+            thd = threading.Thread(target=self.run_job_queue, args=(queue_key,))
```
Added a TODO for the next round.

stanley-cheung commented on observability/test/o11y_tests_manager.py:
```diff
@@ -0,0 +1,306 @@
+
+    def start_all_job_queues(self) -> Dict[QueueKey, threading.Thread]:
+        threads: Dict[QueueKey, threading.Thread] = {}
+        max_job_queue_size = 0
```
Implemented.

stanley-cheung commented on observability/test/o11y_tests_manager.py:
```diff
@@ -0,0 +1,306 @@
+        self.exit_status = 0
+
+    def set_exit_status(self, status: int) -> None:
+        self.exit_status = status
```
Implemented.

stanley-cheung commented on observability/test/o11y_tests_manager.py:
```diff
@@ -0,0 +1,306 @@
+        logger.info('Max job queue size: %d' % max_job_queue_size)
+        return threads
+
+    def run_job_queue(self, queue_key: QueueKey) -> None:
```
Implemented.

stanley-cheung commented on observability/test/o11y_tests_manager.py:
```diff
@@ -0,0 +1,306 @@
+        job_queue = self.job_queues[queue_key]
+        for job in job_queue.jobs:
+            try:
+                proc = self.start_process_in_background(
```
Yes I am aware of this issue and will address that in the next round.

stanley-cheung commented on observability/test/o11y_tests_manager.py:
```diff
@@ -0,0 +1,306 @@
+                self.args.language, 'failed' if self.exit_status else 'passed'))
+            sys.exit(self.exit_status)
+
+class TestManager:
```
Commented noted. Will try to address that in the next round.

stanley-cheung commented on observability/test/run_o11y_tests.py:
```diff
@@ -0,0 +1,1174 @@
+
+LANGUAGE_CONSTANTS = {
+    SupportedLangEnum.GO: LanguageConstants(
+        repo_name = 'grpc-go',
```
Will look into it. Thanks for the pointer.

stanley-cheung commented on observability/test/run_o11y_tests.py:
```diff
@@ -0,0 +1,1174 @@
+        server_start_args = '--port {port}',
+        client_action_cmd = 'interop/observability/client/client',
+        client_action_args = \
+        '--server_host={server_host} ' +
```
Yes, will do as part of the bigger documentation effort.

stanley-cheung commented on observability/test/run_o11y_tests.py:
```diff
@@ -0,0 +1,1174 @@
+        trace_results.test_span_custom_labels(CloudTraceInterface.SENT_SPAN_PREFIX, CLIENT_CUSTOM_LABEL)
+
+# Main
+test_runner = TestRunner(parse_args())
```
Implemented.

stanley-cheung commented on observability/test/run_o11y_tests.py:
```diff
@@ -0,0 +1,1174 @@
+                                      self.args.test_case)
+        self.exit_status = 0
+
+    def set_exit_status(self, status: int) -> None:
```
Implemented.

stanley-cheung commented on observability/test/run_o11y_tests.py:
```diff
@@ -0,0 +1,1174 @@
+
+    def run(self) -> None:
+        try:
+            impl = TestCaseImpl(self)
```
Noted. Added a TODO to try to improve next.

stanley-cheung commented on observability/test/test_utils.py:
```diff
@@ -0,0 +1,177 @@
+    def __str__(self) -> str:
+        return self.value
+
+class ObservabilityConfig:
```
Noted. I kept it for now. Maybe eventually I only need the type as an alias.

stanley-cheung commented on observability/test/run_o11y_tests.py:
```diff
@@ -0,0 +1,1174 @@
+        self.sponge_log_out = open(sponge_log_file, 'a')
+
+    def enable_server_logging(self, method_filters: Optional[List] = None) -> None:
+        self.server_config.setdefault('cloud_logging', {})
```
Noted. This area is a bit of a mess right now. Will try to improve.

stanley-cheung commented on observability/test/run_o11y_tests.py:
```diff
@@ -0,0 +1,1174 @@
+            client_env = self.get_env_with_config(
+                self.client_config,
+                config_file_name = 'client-config-%s.json' % self.identifier)
+        else:
```
Implemented.

stanley-cheung commented on observability/test/run_o11y_tests.py:
```diff
@@ -0,0 +1,1174 @@
+
+    def get_env_with_config(self,
+                            config: ObservabilityConfig,
+                            config_file_name: str = '') -> Dict[str, Any]:
```
Implemented.

stanley-cheung commented on observability/test/run_o11y_tests.py:
```diff
@@ -0,0 +1,1174 @@
+        logger.info('Wait %d seconds before querying cloud...' % WAIT_SECS_READY)
+        time.sleep(WAIT_SECS_READY)
+
+    def get_env_with_config(self,
```
Fixed. Now they are `@staticmethod`.

stanley-cheung commented on observability/test/run_o11y_tests.py:
```diff
@@ -0,0 +1,1174 @@
+        if client_config_into_env_var:
+            client_env[CONFIG_ENV_VAR_NAME] = client_config_into_env_var.toJson()
+        server_proc = self.start_server_in_subprocess(env = server_env)
+        time.sleep(WAIT_SECS_SERVER_START)
```
Noted. Yes this has been on my TODO list from a while ago.

stanley-cheung commented on observability/test/run_o11y_tests.py:
```diff
@@ -0,0 +1,1174 @@
+    def test_event_type_at_least_one(self) -> None:
+        for event_type in self.EVENT_TYPES:
+            logger.debug('testing logging event %s' % event_type)
+            self.assertGreater(self.count_with_type(event_type), 0)
```
Noted. Will address in the next round.

stanley-cheung commented on observability/test/run_o11y_tests.py:
```diff
@@ -0,0 +1,1174 @@
+        interval = monitoring_v3.TimeInterval({
+            'end_time': {'seconds': int(time.time()),
+                         'nanos': test_impl.test_run_metadata.nanos},
+            'start_time': {'seconds': test_impl.test_run_metadata.script_start_seconds,
```
Fixed.

stanley-cheung commented on observability/test/run_o11y_tests.py:
```diff
@@ -0,0 +1,1174 @@
+        })
+        metrics_results: Dict[str, List[ListTimeSeriesResponse]] = {}
+        for metric_name in SUPPORTED_METRICS:
+            method_label = 'grpc_client_method' if 'client' in metric_name else 'grpc_server_method'
```
Fixed.

stanley-cheung commented on observability/test/run_o11y_tests.py:
```diff
@@ -0,0 +1,1174 @@
+        self.assertGreater(num_metrics_result, 0)
+
+    def test_metric_resource_type(self, metric_name: str, resource_type: str) -> None:
+        if self.count_total(metric_name) == 0:
```
Removed.

stanley-cheung commented on observability/test/run_o11y_tests.py:
```diff
@@ -0,0 +1,1174 @@
+            self.assertEqual(result.resource.type, resource_type)
+
+    def test_metric_resource_labels(self, metric_name: str, resource_labels: Dict[str, str]) -> None:
+        if self.count_total(metric_name) == 0:
```
Removed.

stanley-cheung commented on observability/test/run_o11y_tests.py:
```diff
@@ -0,0 +1,1174 @@
+            return
+        for result in self.results[metric_name]:
+            logger.debug('Metric resource labels: %s' % str(result.resource.labels.items()))
+            actual_labels = self.copy_to_dict(result.resource.labels)
```
I think I did this because some of the fields returned by these cloud API calls are protobuf objects, which makes some of those deep comparisons I was trying to do a bit funky. I added a TODO to try to fix this in the next round.

stanley-cheung commented on observability/test/run_o11y_tests.py:
```diff
@@ -0,0 +1,1174 @@
+
+    def enable_client_trace(self, trace_config: Optional[Dict[str, Any]] = None) -> None:
+        self.client_config.setdefault('cloud_trace', {})
+        default = {
```
Will try to clean this area up later.

stanley-cheung commented on observability/test/run_o11y_tests.py:
```diff
@@ -0,0 +1,1174 @@
+                'project_id': PROJECT,
+            })
+
+    def test_trace_basic(self) -> None:
```
Yes post GA.

stanley-cheung commented:
@gnossen Thanks again for the detailed review. I have addressed most of the comments in these first round. 

The biggest changes are that now individual language's interop client/server are now contained in docker images. This main test driver only needs to run those docker images via `docker run` and supply a uniform set of parameters. 

I have left quite some comments as TODO's for the next round of changes because this PR is just too big. I do like to get this merged first and then work on incrementally fixing those issues later. 

As it stands, this main test driver, together with each language's PR (c++, Java and Golang) do have all tests passed for all 3 languages.

gnossen commented on build/kokoro/build.sh:
```diff
@@ -85,72 +87,61 @@ export REPOS_BASE_DIR=${KOKORO_ARTIFACTS_DIR}/github
     git checkout o11y-testing-${JOB_MODE} && \
     git log -1 --oneline && \
-    ./gradlew installDist -x test -PskipCodegen=true -PskipAndroid=true)
+    ./buildscripts/observability-test/build_docker.sh && \
```
Where is `build_docker.sh`? I don't see it in this PR or on the master branch and I don't see it being programmatically generated anywhere? Maybe accidentally not checked in?

stanley-cheung replied:
So this belongs to each individual language's repo.

gnossen reviewed (approved):
Please try to get unresolved comments into TODOs with associated Github issues.

sergiitk commented:
I'll give this another look later this week

stanley-cheung commented:
> Please try to get unresolved comments into TODOs with associated Github issues.

Done. Logged all issues.

stanley-cheung commented:
> I'll give this another look later this week

Thanks! I will probably go ahead and merge this for now and there are still tons of improvement to do.
